rados/read_op_omap_get_vals_by_keys.go

```diff
@@ -0,0 +1,126 @@
+//go:build ceph_preview
+// +build ceph_preview
+
+package rados
+
+// #cgo LDFLAGS: -lrados
+// #include <rados/librados.h>
+// #include <stdlib.h>
+//
+import "C"
+
+import (
+	"unsafe"
+
+	"github.com/ceph/go-ceph/internal/cutil"
+)
+
+// ReadOpOmapGetValsByKeysStep holds the result of the
+// GetOmapValuesByKeys read operation.
+// Result is valid only after Operate() was called.
+type ReadOpOmapGetValsByKeysStep struct {
+	withRefs
+
+	// C arguments
+
+	cKeys cutil.CPtrCSlice
+	iter  C.rados_omap_iter_t
+	prval *C.int
+
+	// Internal state
+
+	// canIterate is only set after the operation is performed and is
+	// intended to prevent premature fetching of data
+	canIterate bool
+}
+
+func newReadOpOmapGetValsByKeysStep(keys []string) *ReadOpOmapGetValsByKeysStep {
+	s := &ReadOpOmapGetValsByKeysStep{
+		cKeys: cutil.NewCPtrCSlice(len(keys)),
+		prval: (*C.int)(C.malloc(C.sizeof_int)),
+	}
+
+	for i, key := range keys {
+		cKey := C.CString(key)
+		s.add(unsafe.Pointer(cKey))
+		s.cKeys[i] = cutil.CPtr(cKey)
+	}
+
+	return s
+}
+
+func (s *ReadOpOmapGetValsByKeysStep) free() {
+	s.canIterate = false
+	C.rados_omap_get_end(s.iter)
+
+	s.cKeys.Free()
+	s.cKeys = nil
+
+	C.free(unsafe.Pointer(s.prval))
+	s.prval = nil
+
+	s.withRefs.free()
+}
+
+func (s *ReadOpOmapGetValsByKeysStep) update() error {
+	err := getError(*s.prval)
+	s.canIterate = (err == nil)
+
+	return err
+}
+
+// Next gets the next omap key/value pair referenced by
+// ReadOpOmapGetValsByKeysStep's internal iterator.
+// If there are no more elements to retrieve, (nil, nil) is returned.
+// May be called only after Operate() finished.
+//  PREVIEW
+func (s *ReadOpOmapGetValsByKeysStep) Next() (*OmapKeyValue, error) {
+	if !s.canIterate {
+		return nil, ErrOperationIncomplete
+	}
+
+	var (
+		cKey    *C.char
+		cVal    *C.char
+		cValLen C.size_t
+	)
+
+	ret := C.rados_omap_get_next(s.iter, &cKey, &cVal, &cValLen)
+	if ret != 0 {
+		return nil, getError(ret)
+	}
+
+	if cKey == nil {
+		// Iterator has reached the end of the list.
+		return nil, nil
+	}
+
+	return &OmapKeyValue{
+		Key:   C.GoString(cKey),
+		Value: C.GoBytes(unsafe.Pointer(cVal), C.int(cValLen)),
+	}, nil
+}
+
+// GetOmapValuesByKeys starts iterating over specific key/value pairs.
+//  PREVIEW
+//
+// Implements:
+//  void rados_read_op_omap_get_vals_by_keys(rados_read_op_t read_op,
+//                                           char const * const * keys,
+//                                           size_t keys_len,
+//                                           rados_omap_iter_t * iter,
+//                                           int * prval)
+func (r *ReadOp) GetOmapValuesByKeys(keys []string) *ReadOpOmapGetValsByKeysStep {
+	s := newReadOpOmapGetValsByKeysStep(keys)
+	r.steps = append(r.steps, s)
+
+	C.rados_read_op_omap_get_vals_by_keys(
+		r.op,
+		(**C.char)(s.cKeys.Ptr()),
+		C.size_t(len(keys)),
+		&s.iter,
+		s.prval,
+	)
+
```

Comment:
Actually, here is the same situation as in #630, the arguments (list of keys) are copied as expected, see [here](https://github.com/ceph/ceph/blob/master/src/librados/librados_c.cc#L4435). You can simplify the code and just free the list strings right after the call, which will reduce the risk of memory leaks.